### PR TITLE
Return list of LayerDataTuple instead of single layers

### PR DESCRIPTION
**Pull request by zoccoler**
I implemented these changes to try to solve #6, returning several layers for several drawn shapes, but it is still not working.

The motivation for this is explained [here](https://napari.org/guides/stable/magicgui.html?highlight=layerdatatuple#returning-list-napari-types-layerdatatuple) and there is indication that this approach has been implemented [here](https://forum.image.sc/t/problem-with-adding-multiple-layers-at-once-as-list-napari-types-layerdatatuple-to/61293).

Does any of you guys have ideas to make this work? @haesleinhuepf @tdmorello

### CODE REVIEW

**Comment by tdmorello** [comment]
Hm, maybe it needs to be decorated with `@magic_factory` for that type annotation to work? I'm not sure.

**Comment by tdmorello** [comment]
The label functionality was broken after my last PR (sorry about that!). I fixed that and added a test for it. If the output type of the crop function changes with @zoccoler 's PR, the tests will need to change. My fork makes these changes (https://github.com/tdmorello/napari-crop) but also implements @zoccoler 's multi crop idea. Not sure what you want to do @haesleinhuepf , but I can hold off on that PR until @zoccoler 's changes go through!

**Comment by haesleinhuepf** [comment]
Hi Tim and Marcelo @tdmorello @zoccoler ,

bug fixes have higher priority than new features. Would you mind sending a PR with your bugfix @tdmorello ? I'd merge it right away. We can then afterwards find out how to incorporate @zoccoler's suggestion (which is great btw). There was a discussion recently in the image.sc forum about a similar issue:
https://forum.image.sc/t/problem-with-adding-multiple-layers-at-once-as-list-napari-types-layerdatatuple-to/61293

Maybe that helps :-)

Thanks!
Robert

**Comment by zoccoler** [comment]
Hi Tim and Robert,

@tdmorello no problem, if #9 from your fork fix both things, we should merge there :) 
Does that implementation actually work for you? I'm refering to the multi-crop functionality by returning a list of `LayerDataTuple`. I guess it should work, but it didn't for me yet.

@haesleinhuepf that's the same link I mentioned 😅 To my best knowledge, I made the same kind of implementation, but it still returns just a single layer.

**Comment by zoccoler** [comment]
The way I had done before was to add the layer inside the function before returning, that should work, but I think using annotations seems more appropriate indeed.

File: napari_crop/_function.py [hunk 1]
```diff
@@ -35,6 +36,7 @@ def crop_region(
 
     shape_types = shapes_layer.shape_type
     shapes = shapes_layer.data
+    cropped_list = []
     for shape, shape_type in zip(shapes, shape_types):
```
Comment (haesleinhuepf) [inline napari_crop/_function.py:40]:
```suggestion
    for count, [shape, shape_type] in enumerate(zip(shapes, shape_types)):
```

File: napari_crop/_function.py [hunk 2]
```diff
@@ -74,9 +76,9 @@ def crop_region(
...
-    layer_props["name"] = layer_props["name"] + " (cropped)"
-    if layer_type == "image":
-        return napari.layers.Image(cropped_data, **layer_props)
-    if layer_type == "labels":
-        return napari.layers.Labels(cropped_data, **layer_props)
+        layer_props["name"] = layer_props["name"] + " (cropped)"
+        if layer_type == "image":
+            cropped_list.append((cropped_data, layer_props, 'image'))
+        if layer_type == "labels":
+            cropped_list.append((cropped_data, layer_props, 'labels'))
```
Comment (haesleinhuepf) [inline napari_crop/_function.py:83]:
```suggestion
        new_layer_props = layer_props.copy()
        new_layer_props["name"] = layer_props["name"] + f" (cropped {count+1})"
        cropped_list.append((cropped_data, new_layer_props, layer_type))
```
It turns out, we cannot pass the same identical dictionary to all new layers. It must be copies. The underlying reason is: if there are two layers with the same name, the data will be stored to the same layer. That's why I was adding a number to the name.
Furthermore, the two `if`-statements were not necessary. We can pass the type-string as parameter as it is.

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
I think I made it to work. We just need to provide different names for each copied layer.

Let me know if my suggestions work for you Marcelo @zoccoler

File: napari_crop/_function.py (hunk 2 above)
Comment (zoccoler) [reply napari_crop/_function.py:83]: Nice @haesleinhuepf ! It works here!
I am going to commit changes and, if that's OK, merge it afterwards.

**Comment by zoccoler** [comment]
I can't merge because I don't have write access to the repository. @haesleinhuepf

**Review by haesleinhuepf — APPROVED** [review]

**Comment by haesleinhuepf** [comment]
Hi Marcelo @zoccoler ,

thanks for implementing this. That was actually some feature I was thinking of from the beginning and never found time for implementing it. I'm glad you did it, :-)

Would you mind checking why the tests are failing at the moment? You can run them locally using `pytest`  and read some details about napari plugin testing [here](https://napari.org/plugins/stable/best_practices.html#write-extensive-tests-for-your-plugin). Furthermore, would you please add a simple test for the new feature? I'd say a test where two squares are cropped and afterwards, it's tested if the resulting layer list contains two elements would be enough.

I will then merge it happily :-)

Thanks!
Robert

**Comment by tdmorello** [comment]
I think it's failing because of the new output type of ``crop_region``.  I'll PR my test file with Marcelo's @zoccoler (merge in a merge... inception!)

**Comment by zoccoler** [comment]
Thank you Robert @haesleinhuepf for these kind words ! :-)
Don't worry Tim @tdmorello, I will put your tests back and edit them to fit the new type of output 😃

**Comment by zoccoler** [comment]
Hmm except for the ellipse, tests work locally, but not here in Github. Do you guys know what I am missing?

About the ellipse error, it fails due to unmatching mask and cropped data shapes. This also happens with polygons drawn close to the image border (`.to_masks()` generates larger (+1px in x/y) masks in these cases). This is a bug, I will open a new issue to deal with that separatedly from here.

**Comment by tdmorello** [comment]
These changes fix `test_crop_function_nd`:

```
diff --git a/napari_crop/_tests/test_function.py b/napari_crop/_tests/test_function.py
index 5522eb0..2f28997 100644
--- a/napari_crop/_tests/test_function.py
+++ b/napari_crop/_tests/test_function.py
@@ -97,7 +97,7 @@ shape_data_ids = ["2x3_crop", "neg_crop", "big_crop", "poly_crop"]
     "shape_data,shape_type", zip(shape_data, shape_types), ids=shape_data_ids
 )
 
-def test_crop_function_nd(image_data, rgb, shape_data, shape_type,
+def test_crop_function_nd(layer_data, rgb, layer_type, shape_data, shape_type,
                           make_napari_viewer):
 
     viewer = make_napari_viewer()
@@ -117,7 +117,7 @@ def test_crop_function_nd(image_data, rgb, shape_data, shape_type,
     nlayers = len(viewer.layers)
 
     #  Get first tuple element (data) of first list element (LayerDataTuple)
-    cropped_data = crop_region(img_layer, shp_layer)[0][0]
+    cropped_data = crop_region(layer, shapes_layer)[0][0]
     viewer.add_image(cropped_data)
 
     assert len(viewer.layers) == nlayers + 1
```

Haven't looked into the ellipse param for `test_crop_function_values_2d` yet.

**Comment by zoccoler** [comment]
> These changes fix `test_crop_function_nd`:

Thanks Tim! @tdmorello I have no idea how that got past 😅

**Comment by zoccoler** [comment]
Hi guys,

I implemented some changes that I think fix #13 and #8 (also #6, which was already fixed before this last commit).

Basically I mapped the output from `.to_masks()` to the cropped data. Tests worked locally.
I also made irregular shape crops to be minimal by trimming zeros.

@haesleinhuepf please let me know if it works. If it does, from my side it is OK to merge.

**Comment by haesleinhuepf** [comment]
Hi Marcelo @zoccoler ,

there is still one test failing (see below). It also fails locally on my  machine using `pytest --cov=napari_crop` from the root directory. Can you take another look please?

```
________________________________________ test_crop_function_values_2d[polygon] ________________________________________

make_napari_viewer = <function make_napari_viewer.<locals>.actual_factory at 0x000002392DD9C940>
shape = array([[0, 2],
       [4, 4],
       [4, 2],
       [2, 0]]), shape_type = 'polygon'
crop_expected = array([[ 0,  7,  0],
       [11, 12, 13],
       [16, 17, 18],
       [ 0, 22,  0]])

    @pytest.mark.parametrize(
        "shape,shape_type,crop_expected",
        zip(shapes, shape_types, crop_expected),
        ids=shape_types,
    )
    def test_crop_function_values_2d(make_napari_viewer, shape, shape_type,
                                     crop_expected):
        """Test that the cropped output is the expected array."""

        viewer = make_napari_viewer()
        img_layer = viewer.add_image(arr_2d)
        shapes_layer = viewer.add_shapes(shape, shape_type=shape_type)
        cropped_actual = crop_region(img_layer, shapes_layer)
        cropped_actual_arrays = [cropped[0] for cropped in cropped_actual][0]
>       assert np.array_equal(crop_expected, cropped_actual_arrays)
E       assert False
E        +  where False = <function array_equal at 0x0000023924978940>(array([[ 0,  7,  0],\n       [11, 12, 13],\n       [16, 17, 18],\n       [ 0, 22,  0]]), array([[ 0,  2,  0],\n       [ 6,  7,  0],\n       [11, 12, 13],\n       [ 0, 17, 18]]))
E        +    where <function array_equal at 0x0000023924978940> = np.array_equal

napari_crop\_tests\test_function.py:61: AssertionError
```

**Comment by zoccoler** [comment]
Ok, could you please try again now?

**Comment by haesleinhuepf** [comment]
Big thanks @zoccoler and @tdmorello :-)

**Comment by zoccoler** [comment]
Close #13

**Pull request merged**